Recategorizing built-in methods

yeexinc opened the pull request:
### Purpose

This PR is to address [DYN-421](https://jira.autodesk.com/browse/DYN-421).
The built-in methods are recategorized to the following functions: (ticked ones are the ones committed in this PR)
- [x] ImportExport.ImportCSV
- [x] List.AllFalse
- [x] List.AllTrue
- [x] List.Contains
- [x] List.IsHomogeneous
- [x] List.IsRectangular
- [x] List.IsUniformDepth
- [ ] List.TrueForAll
- [ ] List.TrueForAny
- [x] List.SetDifference
- [x] List.SetIntersection
- [x] List.SetUnion
- [x] List.IndexOf
- [ ] List.Equals
- [x] List.CountFalse
- [x] List.CountTrue
- [x] List.Count
- [ ] List.RemoveIfNot
- [x] List.Flatten
- [x] List.Insert
- [ ] List.GroupByFunction
- [x] List.Reorder
- [ ] List.SortByFunction
- [ ] List.MaximumItemByKey
- [ ] List.MinimumItemByKey
- [ ] List.ContainsKeys
- [ ] List.RemoveKey
- [x] List.SortIndexByValue
- [x] List.NormalizeDepth
- [ ] List.GetValues
- [x] Math.Map
- [x] Math.MapTo
- [x] Math.PI
- [ ] Math.NewtonRootFind1DNoDeriv
- [ ] Math.NewtonRootFind1DWithDeriv

To be consistent with the migration of `BuiltIn.ImportFromCSV` to `ImportExport.ImportCSV`, the following method is added as well:
- [x] ImportExport.ExportCSV (to replace CSV.WriteToFile)

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@aparajit-pratap 

### FYIs

@ke-yu @sharadkjaiswal

gregmarr commented on src/Libraries/CoreNodes/List.cs:
```diff
@@ -52,6 +52,362 @@ public static bool ContainsItem(IList list, object item)
+        /// <search>flatten,1D</search>
+        public static IList Flatten(IList list)
+        {
+            return Flatten(list, GetDepth(list), new List<object>());
```
I initially mapped builtin Flatten to List.Flatten in Thunderstorm.  Unfortunately, this results in a change in output for the case where builtin Flatten is passed a simple value rather than a list.  It currently returns the value unmodified rather than putting it in a list.

yeexinc replied:
Thanks for the review @gregmarr , I'll look into that.

gregmarr replied:
@yeexinc I was updating the data from the tests in the CoGS repo, and found that this output change has happened, demonstrated in `testPlattenCompletely_singleInput.dyn`.  Looks like the new `List.Flatten` with no argument isn't behaving the same as the builtin `Flatten`.

yeexinc commented:
The current state of this PR is:
- Built-in methods that have been re-implemented (the ticked items in the list) are hidden from user's search
- Icons have been moved from `BuiltInImages.resx` to `DSCoreNodesImages.resx`
- Test cases have been added for the new methods added

gregmarr commented:
@yeexinc Can you look at this code and see if it lines up with what's been done thus far?
https://git.autodesk.com/AEC-Generative-Design/ComputationGraphService/blob/master/src/codegen/codeblocknode.go#L579-L619

yeexinc commented:
Hi @gregmarr ,
- The nodes with `Autodesk.DesignScript.Geometry` in their names have not yet been recategorized.
- `List.FlattenCompletely` is named as `List.Flatten` with two overloads, `Flatten(list)` and `Flatten(list, amt)`, in which the first function will flatten an input list completely.
- `Concat` is not in the BuiltIn methods - it seems to be replaced by `Join`?
- `List.ContainsItem` checks if the item is present in the values of the list, and doesn't function the same way as `ContainsKey`.

The others seem to line up excluding the ones unchecked in the description.

gregmarr commented:
So `List.Flatten` now has two overloads?

"Concat is not in the BuiltIn methods - it seems to be replaced by Join?"
https://git.autodesk.com/Dynamo/Dynamo/blob/master/src/Engine/ProtoCore/Lang/BuiltInMethods.cs#L204-L215

"doesn't function the same way as ContainsKey."

Yeah, I'm not sure what we should do with that one.  TS doesn't support lists as dictionaries.

gregmarr commented:
Note that the spreadsheet has `List.ContainsKeys from BuiltIn > ContainsKeys`, plural rather than the current singular.

yeexinc commented:
@gregmarr - Thanks for the input, the `Concat` function is hidden by setting `HiddenInLibrary` to true:
https://git.autodesk.com/Dynamo/Dynamo/blob/master/src/Engine/ProtoCore/Parser/AssociativeAST.cs#L1541-L1543

Current implementation for `List.Flatten` has two overloads:
![list flatten](https://cloud.githubusercontent.com/assets/17231814/23597152/7334387e-026b-11e7-90aa-71ac3239bfc4.png)

gregmarr commented:
Instead of adding an overload, should it have been added as a new name?

gregmarr commented:
At least for now, I need to handle all possible builtins, until the Dynamo migration eliminates them all.

aparajit-pratap commented on src/Libraries/CoreNodes/List.cs:
```diff
@@ -52,6 +52,375 @@ public static bool ContainsItem(IList list, object item)
+        /// <param name="indice">The indices used to reorder the items in the list.</param>
+        /// <returns name="list">The reordered list.</returns>
+        /// <search>reorder,index,indices</search>
+        public static IList Reorder(IList list, IList indice)
```
Hidden from built-in?

aparajit-pratap commented on src/Libraries/CoreNodes/List.cs:
```diff
@@ -52,6 +52,375 @@ public static bool ContainsItem(IList list, object item)
+        /// </summary>
+        /// <param name="list">The list whose depth is to be normalized.</param>
+        /// <returns name="list">The list with the normalized rank.</returns>
+        public static IList NormalizeDepth(IList list)
```
The two `NormalizeDepth` methods must be consolidated as per the spreadsheet. Maybe you can add the rank as the second parameter with default argument

aparajit-pratap commented on src/Libraries/CoreNodes/List.cs:
```diff
@@ -52,6 +52,375 @@ public static bool ContainsItem(IList list, object item)
+            return list;
+        }
+
+        private static IList IncreaseDepth(IList list, int amt)
```
I think it would be easier to read this file if all private helper methods are grouped together within a region and all public methods are together in a separate region.

aparajit-pratap commented on src/Libraries/CoreNodes/List.cs:
```diff
@@ -52,6 +52,375 @@ public static bool ContainsItem(IList list, object item)
+        /// </summary>
+        /// <param name="list">The list to obtain the values from.</param>
+        /// <returns name="newlist">The new list which contains the values.</returns>
+        public static IList GetValues(IList list)
```
We need to revisit this obviously ...

Racel commented:
@aparajit-pratap @ke-yu - Can you please provide a list of which nodes were problematic to convert to ZT? Also, please provide a reason why they are difficult to implement.

ke-yu commented:
@Racel @kronz @riteshchandawar @aparajit-pratap @Benglin 

It needs runtime information to do deep comparison (e.g., comparing two Points needs to get .net objects and call `Equals()` on these two objects).
- List.Equals

`List.RemoveIfNot()` removes those items that are not of specified type (e.g., "string", "int" etc). It needs runtime information (to get the type of object), so can't be implemented in zero touch library. 
 - List.RemoveIfNot

Partial function. These functions are implemented in pure DesignScript code. We don't have a way to marshal DesignScript function to .net function and run it in .net land. 
- List.GroupByFunction
- List.SortByFunction
- List.MaximumItemByKey
- List.MinimumItemByKey
- List.TrueForAll
- List.TrueForAny

Dictionary related functions, I think we could provide zero touch functions for them @yeexinc could you please try? Though we are not encouraging user to use dictionary. 
- List.RemoveKey
- List.GetValues
- List.ContainsKeys

Beyond human intelligence to understand how it work.
- Math.NewtonRootFind1DNoDeriv
- Math.NewtonRootFind1DWithDeriv

Racel commented:
Thanks for the list @ke-yu. Here are my suggestions:
- Deprecate both Math.NewtonRoot nodes
- Visually place all of the other nodes as list, but just leave them as builtins. Can you tell us what that would look like if you called these remaining built-in nodes in DesignScript? 
- What do people usually use Equals for as opposed to ==? Is it mainly for geometry? If so, maybe we should place Equals in Geometry.Geometry. Thoughts?


I especially enjoyed 
> Beyond human intelligence to understand how it work.

yeexinc commented:
Thanks for the review @aparajit-pratap , I've made some changes:
- Edited ImportExport.ImportFromCSV, the implementation is about the same as `LoadCSVWithMode` in `BuiltInFunctionEndPoints.cs`, since it is the implementation for `BuiltIn.ImportFromCSV`. `DSCore.IO.CSV.ReadFromFile` has a separate implementation.
- Hidden `IsRectangular` and `Reorder` in built-ins.
- Simplified `SetUnion`, `SetInteraction` and `SetDifference` so that they exert the same behaviour has built-ins.
- Renamed `Flatten` (without the amt parameter) to `FlattenCompletely`.
- Added test cases for `ImportFromCSV` and edited some of the other test cases.

@ke-yu - I'm not sure if I can cast the inputs into Dictionary class as they're passed as ArrayList (with the information about their keys lost) into the zero touch functions. Any better ways of implementing it?

aparajit-pratap commented on src/Libraries/CoreNodes/List.cs:
```diff
@@ -40,15 +42,267 @@ public static IList UniqueItems(IList list)
+        }
+
+        /// <summary>
+        ///     Returns multidimentional list according the rank given.
```
multidimensional

gregmarr commented on src/Libraries/CoreNodes/List.cs:
```diff
@@ -40,15 +42,267 @@ public static IList UniqueItems(IList list)
+        /// <param name="list">The list to be flattened.</param>
+        /// <returns name="list">The flattened 1D list.</returns>
+        /// <search>flatten,1D</search>
+        public static IList FlattenCompletely(IList list)
```
Builtin `Flatten()` took no arguments.  When I made the Thunderstorm implementation of this Builtin, I mapped it to a new function `List.FlattenCompletely()` because of the existing `List.Flatten()`, rather than adding an overload.

aparajit-pratap replied:
@gregmarr I believe we are not supporting method overloads in TS, while at the same time we aren't removing them from DS (at least for the time being). So how are we addressing this in Thunderstorm? Does TS uniquely rename the overloads based on their arguments?

aparajit-pratap replied:
I understand from @ke-yu that these methods are named differently in TS. @yeexinc could you rename this method to `Flatten` and make it an overload. cc @Racel

gregmarr replied:
TS has overloads, but why do we want to go back to having an overload here?  It seems unnecessary.

aparajit-pratap replied:
I think this decision comes from a UX standpoint of minimizing the number of different method names that have similar functionality since it is confusing to users. I'll leave @Racel to elaborate on this.

gregmarr replied:
I implemented it as a separate function, because I thought that we wanted to minimize the number of overloads since it is confusing to users.  If that's not the case, and we actually want the opposite, then the overload makes sense.  If we stay with the overload, we'll update the Thunderstorm version to match.

yeexinc commented on src/Libraries/CoreNodes/Math.cs:
```diff
@@ -81,6 +81,50 @@ public static double Average(IList<double> numbers)
+            }
+            if (inputValue >= rangeMax) return targetRangeMax;
+            else if (inputValue <= rangeMin) return targetRangeMin;
+            return ((inputValue - rangeMin) / (rangeMax - rangeMin)) * (targetRangeMax - targetRangeMin) + targetRangeMin;
```
I wasn't sure how `BuiltIn` handles this. I'll edit these `Map` and `MapTo` function to reuse the implementations in `BuiltIn`.

yeexinc commented:
@gregmarr @Racel @aparajit-pratap
Since according to the original specification, `BuiltIn.Flatten` was to be recategorized as `List.Flatten`, I renamed `List.FlattenCompletely` back to `List.Flatten` so that it becomes an overload. If the name is to be changed again, I think we can have a separate PR for it. 
Thanks.

aparajit-pratap commented on src/Libraries/CoreNodes/ImportExport.cs:
```diff
@@ -0,0 +1,117 @@
+    /// <summary>
+    ///     Methods for Import/Export category.
+    /// </summary>
+    public static class ImportExport
```
As per the recategorization doc https://docs.google.com/spreadsheets/d/1IkmjLemPHfPhCfSW75R3NsUwlFkHpcVMDyvI8WvTiS0/edit#gid=0, this class should be `Data`. There will be no class called `ImportExport`. `ImportExport` will just be a category. I'm merging this in for now. @yeexinc please make a note of this and refactor into separate classes once this is merged into the `LibraryReorg` branch.

yeexinc replied:
Okay. Thank you very much for the review!

Racel commented:
@gregmarr @yeexinc @aparajit-pratap - I responded on slack to the flatten issue. But, as outlined in the recategorization spreadsheet, I suggested to consolidate the builtin and list "flatten" node into one. What do you think about this? We would get rid of the overload and reduce the number of OOTB nodes.

Users can still get the same builtin result but using the List node. For the List.Flatten node, we can set the "amt" input to default to 1. Let me know your thoughts.

gregmarr commented:
This was already done.

yeexinc commented:
@Racel - I thought it'd be good to mention that the PR sets the default value of "amt" to -1, which will flatten a given list completely. Setting the "amt" to 1 will flatten the list only by one level.

Racel commented:
@yeexinc - ah, gotcha. Good point. Can you please add a note about this in the tooltip for the amt input and in the tooltip for the node? Thanks.
